### PR TITLE
feat: add openGraph image to public insight pages

### LINKED ISSUE #686: Feature: Share button
### Type of feature

🍕 Feature

### Current behavior

A user needs to copy the link the URL bar to share their insights page. 

### Suggested solution

The share button should show if the insights pages are marked as public. The functionality should copy the link to the clipboard for now (toast notification). In the future, we should open a settings window with options. 

This will also need a social card.

- [x] #687

Insight Page - Repositories - Period v04

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [ ] I agree to follow this project's Contribution Docs

### LINKED ISSUE #687: Feature: Social Card for shareable insights
When sharing an insights page on social, we need to generate a social card on demand. Similar to how GitHub powered to social cards. 

https://github.blog/2021-06-22-framework-building-open-graph-images/

## Figma
Cards needed
- [x] Profile (Share button should get added to profile)
- [ ] Insight Pages (related https://github.com/open-sauced/insights/issues/686)
- [x] Highlights (waiting on https://github.com/open-sauced/insights/pull/859)

https://www.figma.com/file/OpVX6WT7dmWqnwRuEvADMF/OpenSauced-%E2%80%94-Insights-(insights.opensauced.pizza)?node-id=6756%3A8597&t=PKXNvKud1Q5bXoAU-0

### PR DESCRIPTION
**Pull request by OgDev-01**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description
Add SEO component to the HubPage component and fetch social card URL. If the page is not hydrated, return only the SEO component. If the page is hydrated, return the Tool component and the SEO component. The social card URL is fetched to display the image on social media platforms.

_Generated using [OpenSauced](https://opensauced.ai/)._

## Related Tickets & Documents
Closes #687 
Closes #686 

### Desktop
image

### Mobile
image

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

### CODE REVIEW

File: components/molecules/InsightHeader/insight-header.tsx [hunk 1]
```diff
@@ -8,6 +11,8 @@ import ContextThumbnail from "components/atoms/ContextThumbnail/context-thumbnai
 import CardRepoList from "../CardRepoList/card-repo-list";
 import { truncateString } from "lib/utils/truncate-string";
 import useRepositories from "lib/hooks/api/useRepositories";
+import { useEffect, useState } from "react";
```
Comment (brandonroberts) [inline components/molecules/InsightHeader/insight-header.tsx:14]: nit: move to top of file

```suggestion
import { useEffect, useState } from "react";
```

File: pages/pages/[userOrg]/[filterName]/[toolName].tsx [hunk 2]
```diff
@@ -24,7 +30,30 @@ const HubPage: WithPageLayout<InsightPageProps> = ({ insight, pageName }: Insigh
...
+          twitterCard="summary_large_image"
+        />
+      </>
+    );
+  }
+
+  return (
+    <>
+      <SEO
+        title={`${insight.name} | Open Sauced Insights Hub`}
```
Comment (brandonroberts) [inline pages/pages/[userOrg]/[filterName]/[toolName].tsx:49]:
```suggestion
        title={`${insight.name} | Open Sauced Insights`}
```

File: pages/pages/[userOrg]/[filterName]/[toolName].tsx [hunk 3]
```diff
@@ -24,7 +30,30 @@ const HubPage: WithPageLayout<InsightPageProps> = ({ insight, pageName }: Insigh
...
+        description={`${insight.name} Insights page on OpenSauced`}
```
Comment (brandonroberts) [inline pages/pages/[userOrg]/[filterName]/[toolName].tsx:50]:
```suggestion
        description={`${insight.name} Insights on OpenSauced`}
```

File: pages/pages/[userOrg]/[filterName]/[toolName].tsx [hunk 4]
```diff
@@ -58,10 +88,31 @@ export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
     };
   }
 
+  // Keeping this here so we are sure the page is not private before we fetch the social card.
+  async function fetchSocialCardURL() {
```
Comment (brandonroberts) [inline pages/pages/[userOrg]/[filterName]/[toolName].tsx:92]: Turn this into a reusable utility function and put it in `lib/utils`. The only thing that changes is the API path for the user/highlight/insight. We can refactor those in a separate PR.

Example code:

```ts
const ogImage = await fetchSocialCardURL(`/insights/${insightId}`); // returns image URL or empty string
```

**Review by brandonroberts — CHANGES_REQUESTED** [review]

File: pages/user/[username]/index.tsx [hunk 5]
```diff
@@ -111,20 +112,10 @@ export async function handleUserSSR({ params }: GetServerSidePropsContext<{ user
...
-        method: "HEAD",
-      }); // trigger the generation of the social card
-    }
-
-    ogImage = ogReq.headers.get("x-amz-meta-location");
-  }
-
   // Runs the data fetching in parallel. Decreases the loading time by 50%.
-  await Promise.allSettled([fetchUserData(), fetchSocialCardURL()]);
+  const [, ogData] = await Promise.allSettled([fetchUserData(), fetchSocialCard(`users/${username}`)]);
```
Comment (bdougie) [inline pages/user/[username]/index.tsx:116]: Maybe naive question, but the empty destructured item on the left a normal practice for variables or is this an oversight?
```
  const [, ogData] =...

```

**Review by brandonroberts — APPROVED** [review]

**Review by bdougie — APPROVED** [review]

File: pages/user/[username]/index.tsx (hunk 5 above)
Comment (brandonroberts) [reply pages/user/[username]/index.tsx:116]: It's normal practice in the case that you only want the second item in the returned array. For clarity we could just flip the two functions calls.

```ts
const [ogData] = Promise.allSettled([fetchSocialCard(), fetchUserData()]);
```

**Pull request merged**